### PR TITLE
feat/TD-23739: enhance ttl, deletetime modified by writing-op

### PR DESCRIPTION
**Pull request by LiShunGang**
control by enableTtlByCtime (default 0)

### CODE REVIEW

File: source/common/src/tmsg.c [hunk 1]
```diff
@@ -3159,7 +3162,7 @@ int32_t tSerializeSVDropTtlTableReq(void *buf, int32_t bufLen, SVDropTtlTableReq
   tEncoderInit(&encoder, buf, bufLen);
 
   if (tStartEncode(&encoder) < 0) return -1;
-  if (tEncodeI32(&encoder, pReq->timestamp) < 0) return -1;
+  if (tEncodeI64(&encoder, pReq->timestamp) < 0) return -1;
```
Comment (hzcheng) [inline source/common/src/tmsg.c]: Why change from i32 to i64? How can u ensure compatibility?

File: source/common/src/tmsg.c [hunk 2]
```diff
@@ -6908,6 +6910,23 @@ int32_t tDecodeSVAlterTbReq(SDecoder *pDecoder, SVAlterTbReq *pReq) {
     default:
       break;
   }
+  return 0;
+}
+
+int32_t tDecodeSVAlterTbReq(SDecoder *pDecoder, SVAlterTbReq *pReq) {
+  if (tStartDecode(pDecoder) < 0) return -1;
+  if (tDecodeSVAlterTbReqCommon(pDecoder, pReq) < 0) return -1;
+  if (tDecodeI64(pDecoder, &pReq->ctimeMs) < 0) return -1;
```
Comment (hzcheng) [inline source/common/src/tmsg.c]: If you want to ensure compactibility, write code like below:

if (tDecodeIsEnd(pDecoder)) {
  if (tDecodeI64(pDecoder, &pReq->ctimeMs) < 0) return -1;
} else {
  pReq->ctimeMs = 0; /* some default value */
}

File: source/common/src/tmsg.c [hunk 3]
```diff
@@ -7205,6 +7225,7 @@ int32_t tDecodeDeleteRes(SDecoder *pCoder, SDeleteRes *pRes) {
 
   if (tDecodeCStrTo(pCoder, pRes->tableFName) < 0) return -1;
   if (tDecodeCStrTo(pCoder, pRes->tsColName) < 0) return -1;
+  if (tDecodeI64(pCoder, &pRes->ctimeMs) < 0) return -1;
```
Comment (hzcheng) [inline source/common/src/tmsg.c]: Please ensure compatibility.

File: source/common/src/tmsg.c [hunk 4]
```diff
@@ -7445,6 +7467,20 @@ int32_t tDecodeSBatchDeleteReq(SDecoder *pDecoder, SBatchDeleteReq *pReq) {
   return 0;
 }
 
+int32_t tDecodeSBatchDeleteReq(SDecoder *pDecoder, SBatchDeleteReq *pReq) {
+  if (tDecodeSBatchDeleteReqCommon(pDecoder, pReq)) return -1;
+  if (tDecodeI64(pDecoder, &pReq->ctimeMs) < 0) return -1;
```
Comment (hzcheng) [inline source/common/src/tmsg.c]: Compactibility!

File: source/common/src/tmsg.c [hunk 5]
```diff
@@ -7558,6 +7595,10 @@ static int32_t tDecodeSSubmitTbData(SDecoder *pCoder, SSubmitTbData *pSubmitTbDa
       pCoder->pos += (*ppRow)->len;
     }
   }
+  if (tDecodeI64(pCoder, &pSubmitTbData->ctimeMs) < 0) {
+    code = TSDB_CODE_INVALID_MSG;
```
Comment (hzcheng) [inline source/common/src/tmsg.c]: Compatibility!

File: include/common/tglobal.h [hunk 6]
```diff
@@ -119,6 +119,7 @@ extern bool    tsQueryUseNodeAllocator;
 extern bool    tsKeepColumnName;
 extern bool    tsEnableQueryHb;
 extern bool    tsEnableScience;
+extern bool    tsEnableTtlByCtime;
```
Comment (hzcheng) [inline include/common/tglobal.h]: Can u explain the option?

File: source/dnode/vnode/CMakeLists.txt [hunk 7]
```diff
@@ -27,6 +27,7 @@ target_sources(
     "src/meta/metaEntry.c"
     "src/meta/metaSnapshot.c"
     "src/meta/metaCache.c"
+    "src/meta/ttl.c"
```
Comment (hzcheng) [inline source/dnode/vnode/CMakeLists.txt]: Please change ttl.c as metaTtl.c.

File: source/common/src/tmsg.c (hunk 1 above)
Comment (LiShunGang) [reply source/common/src/tmsg.c]: To unify the unit to millisecond. Both ctime and btime use millisecond values.
But we can keep it in seconds for compatibility reasons.

File: source/common/src/tglobal.c [hunk 8]
```diff
@@ -511,6 +512,7 @@ static int32_t taosAddServerCfg(SConfig *pCfg) {
   if (cfgAddInt32(pCfg, "mqRebalanceInterval", tsMqRebalanceInterval, 1, 10000, 1) != 0) return -1;
   if (cfgAddInt32(pCfg, "ttlUnit", tsTtlUnit, 1, 86400 * 365, 1) != 0) return -1;
   if (cfgAddInt32(pCfg, "ttlPushInterval", tsTtlPushInterval, 1, 100000, 1) != 0) return -1;
+  if (cfgAddBool(pCfg, "enableTtlByCtime", tsEnableTtlByCtime, 0) != 0) return -1;
```
Comment (hzcheng) [inline source/common/src/tglobal.c]: Please make the configuration  `enableTtlByCtime` more understandable.

**Review by hzcheng — APPROVED** [review]

File: source/common/src/tmsg.c (hunk 3 above)
Comment (LiShunGang) [reply source/common/src/tmsg.c]: fixed

File: source/common/src/tmsg.c (hunk 2 above)
Comment (LiShunGang) [reply source/common/src/tmsg.c]: fixed

File: source/common/src/tmsg.c (hunk 4 above)
Comment (LiShunGang) [reply source/common/src/tmsg.c]: fixed

File: source/common/src/tmsg.c (hunk 5 above)
Comment (LiShunGang) [reply source/common/src/tmsg.c]: fixed

File: source/dnode/vnode/CMakeLists.txt (hunk 7 above)
Comment (LiShunGang) [reply source/dnode/vnode/CMakeLists.txt]: renamed

File: include/common/tglobal.h (hunk 6 above)
Comment (LiShunGang) [reply include/common/tglobal.h]: explain in tglobal.c

File: source/common/src/tglobal.c (hunk 8 above)
Comment (LiShunGang) [reply source/common/src/tglobal.c]: Renamed `enableTtlByCtime`  -> `ttlChangeOnWrite` by suggestion

**Review by stephenkgu — APPROVED** [review]

**Pull request merged**